File: kamacoder99/kamacoder99.py
# dfs

# direction = [[0, 1], [1, 0], [0, -1], [-1, 0]]

# def dfs(grid, visited, x, y):
#     for i, j in direction:
#         next_x = x + i
#         next_y = y + j

#         if next_x < 0 or next_x >= len(grid) or next_y < 0 or next_y >= len(grid[0]):
#             continue
#         if not visited[next_x][next_y] and grid[next_x][next_y] == 1:
#             visited[next_x][next_y] = True
#             dfs(grid, visited, next_x, next_y)

# if __name__ == '__main__':
#     n, m = map(int, input().split())
#     grid = []
#     for i in range(n):
#         grid.append(list(map(int, input().split())))
#     visited = [[False] * m for _ in range(n)]

#     res = 0

#     for i in range(n):
#         for j in range(m):
#             if grid[i][j] == 1 and not visited[i][j]:
#                 res += 1
#                 visited[i][j] = True
#                 dfs(grid, visited, i, j)

#     print(res)

# bfs

from collections import deque
directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
def bfs(grid, visited, x, y):
    que = deque([])
    que.append([x,y])
    while que:
        cur_x, cur_y = que.popleft()
        for i, j in directions:
            next_x = cur_x + i
            next_y = cur_y + j
            if next_y < 0 or next_x < 0 or next_x >= len(grid) or next_y >= len(grid[0]):
                continue
            if not visited[next_x][next_y] and grid[next_x][next_y] == 1: 
                visited[next_x][next_y] = True
                que.append([next_x, next_y])

def main():
    n, m = map(int, input().split())
    grid = []
    for i in range(n):
        grid.append(list(map(int, input().split())))
    visited = [[False] * m for _ in range(n)]
    res = 0
    for i in range(n):
        for j in range(m):
            if grid[i][j] == 1 and not visited[i][j]:
                res += 1
                bfs(grid, visited, i, j)
    print(res)

if __name__ == "__main__":
    main()

File: kamacoder99/kamacoder99.cpp
// dfs

// #include <iostream>
// #include<bits/stdc++.h>

// using namespace std;

// int dir[4][2] = {0, 1, 1, 0, -1, 0, 0, -1};

// void dfs(const vector<vector<int>>& grid, vector<vector<bool>>& visited, int x, int y){
//     for (int i = 0; i < 4; i++){
//         int nextx = x + dir[i][0];
//         int nexty = y + dir[i][1];

//         if (nextx < 0 || nextx >= grid.size() || nexty < 0 || nexty > grid[0].size()){
//             continue;
//         }

//         if (!visited[nextx][nexty] && grid[nextx][nexty] == 1){
//             visited[nextx][nexty] = true;
//             dfs(grid, visited, nextx, nexty);
//         }
//     }
// }


// int main(){
//     int n, m;
//     cin >> n >> m;
//     vector<vector<int>> grid(n, vector<int>(m, 0));
//     for (int i = 0; i < n; i++){
//         for (int j = 0; j < m; j++){
//             cin >> grid[i][j];
//         }
//     }

//     vector<vector<bool>> visited(n, vector<bool>(m, false));

//     int result = 0;
//     for (int i = 0; i < n; i++){
//         for (int j = 0; j < m; j++){
//             if (!visited[i][j] && grid[i][j] == 1){
//                 visited[i][j] = true;
//                 result ++;
//                 dfs(grid, visited, i, j);
//             }
//         }
//     }
//     cout << result << endl;
// }


// bfs

#include <iostream>
#include <bits/stdc++.h>

using namespace std;

int dir[4][2] = {0, 1, 1, 0, -1, 0, 0, -1}; 
void bfs(const vector<vector<int>>& grid, vector<vector<bool>>& visited, int x, int y) {
    queue<pair<int, int>> que;
    que.push({x, y});
    visited[x][y] = true;
    while(!que.empty()) {
        pair<int ,int> cur = que.front(); que.pop();
        int curx = cur.first;
        int cury = cur.second;
        for (int i = 0; i < 4; i++) {
            int nextx = curx + dir[i][0];
            int nexty = cury + dir[i][1];
            if (nextx < 0 || nextx >= grid.size() || nexty < 0 || nexty >= grid[0].size()) continue;  // 越界了，直接跳过
            if (!visited[nextx][nexty] && grid[nextx][nexty] == 1) {
                que.push({nextx, nexty});
                visited[nextx][nexty] = true; 
            }
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> grid(n, vector<int>(m, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> grid[i][j];
        }
    }

    vector<vector<bool>> visited(n, vector<bool>(m, false));

    int result = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (!visited[i][j] && grid[i][j] == 1) {
                result++; 
                bfs(grid, visited, i, j); 
            }
        }
    }

    cout << result << endl;
}